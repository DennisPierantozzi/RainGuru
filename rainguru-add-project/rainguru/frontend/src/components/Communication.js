import React from "react";
import Popup from "react-popup";
import Map from "./Map/Map";

export default class Communication {
    static useLatestData = true;
    static observed = false;
    static dataCollected = false;
    static imageUrls = [];
    static latestTimestamp = 0;
    static requestTimestamp = [];
    static dataTimestamp = [];

    // getters
    /**
     * Get the boolean representing if we use the latest data.
     */
    static getUseLatestData() {
        return this.useLatestData;
    }

    /**
     * Get the boolean representing if we show observed or predicted data.
     */
    static getObserved() {
        return this.observed;
    }

    /**
     * Get the boolean representing if all the data has been loaded.
     */
    static getDataCollected() {
        return this.dataCollected;
    }

    /**
     * Get the urls for the images with the predicted precipitation.
     */
    static getImageUrls() {
        return this.imageUrls;
    }

    /**
     * Get the timestamp of the first prediction of the latest data.
     */
    static getLatestTimestamp() {
        return this.latestTimestamp;
    }

    /**
     * Get the timestamp that is used when making requests.
     */
    static getRequestTimestamp() {
        return this.requestTimestamp;
    }

    /**
     * Get the timestamp of the first prediction of the data in use.
     */
    static getDataTimestamp() {
        return this.dataTimestamp;
    }

    // setters
    /**
     * Set the boolean representing to if we use the latest data.
     */
    static setUseLatestData(latest) {
        this.useLatestData = latest;
    }

    /**
     * Set the boolean representing if we show observed or predicted data.
     */
    static setObserved(observed) {
        this.observed = observed;
    }

    /**
     * Set the timestamp that the requests will use
     */
    static setRequestTimestamp(timestamp) {
        this.requestTimestamp = timestamp;
    }


    /**
     * Fetches new data from the back-end.
     */
    static async fetchUrls() {
        let url = "api/fetch?observed=" + this.observed;
        // only add timestamp if not fetching latest data
        if (!this.useLatestData) {
            url += "&timestamp=" + this.requestTimestamp;
        }
        // make request to retrieve predictions
        await fetch(url)
          .then(response => {
            if (response.status > 400) {
              throw new Error();
            }
            return response.json();
          })
          .then(responseJson => {
            // store predictions and create images for them
            this.imageUrls = responseJson.urls;
            this.dataTimestamp = new Date(responseJson.timestamp * 1000);
            if(responseJson.exception_active) {
                Popup.alert("Message from server:\n" + responseJson.exception_message, "Warning: Model might be using old data.");
                console.log(responseJson.exception_message);
            }
            Map.preloadedImages();
            this.dataCollected = true;
          });
    }

    /**
     * Fetches the timestamp of the first prediction from the latest data.
     *
     * @returns {Promise<boolean>} representing if there is new data available.
     */
    static async checkNewData() {
        let dataUpdated = false;
        // make a request for the latest data first prediction timestamp
        await fetch("api/check")
            .then(response => {
                if (response.status > 400) {
                  throw new Error();
                }
                return response.json();
            })
            .then(responseJson => {
                // check if there is new data via the timestamp received
                const newTimestamp = responseJson.timestamp;
                if (newTimestamp !== this.latestTimestamp) {
                    this.latestTimestamp = newTimestamp;
                    dataUpdated = true;
                }
            });
        return dataUpdated;
    }

    /**
     * Fetches the data about a selected location for a selected 100 minutes.
     *
     * @param x the x coordinate of the 480x480 images
     * @param y the y coordinate of the 480x480 images
     * @returns {Promise<any>} the precipitation of the location for a selected 100 minutes.
     */
    static async fetchPrecipitation(x, y) {
        let url = "api/precipitation?x=" + x + "&y=" + y + "&observed=" + this.observed;
        // only add timestamp if not fetching latest data
        if (!this.useLatestData) {
            url += "&timestamp=" + this.requestTimestamp;
        }

        // make a request for the precipitation information of the give location
        return await fetch(url)
            .then(response => {
                if (response.status > 400) {
                    throw new Error();
                }
                return response.json();
            })
            .then(responseJson => {
                // return the fetched precipitation
                return responseJson.precipitation;
            })
    }

    /**
     * Fetch the times of what data is all available
     *
     * @returns {Promise<{available: *, predictedTimes: *, observedTimes: *}>} the availability of data times
     */
    static async fetchDataAvailability() {
        // make a request to get the availability times
        return await fetch("api/available")
            .then(response => {
                if (response.status > 400) {
                    throw new Error()
                }
                return response.json();
            })
            .then(responseJson => {
                // return a boolean representing if data is being stored and the available data times
                return {
                    available: responseJson.store_data,
                    predictedTimes: responseJson.predictions_stored,
                    observedTimes: responseJson.observations_stored
                };
            })
    }
}