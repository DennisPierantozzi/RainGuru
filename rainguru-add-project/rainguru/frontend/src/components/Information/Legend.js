import React, { Component } from "react";

export default class Legend extends Component {
    constructor(props) {
        super(props);
    }

    /**
     * Renders the legend.
     * It renders the following:
     *  - Render the top text above the legend.
     *  - Render the legend with some numbers following a logarithmic scale with rain intensity.
     *
     * @returns the rendered legend.
     */
    render() {
        return (
            <div className="legendContents" data-testid="legend-box">
                <div className="legendTitle">Precipitation [mm/h]</div>
                <div className="legendContainer">
                    <div className="legend">
                        <div id="gradient1container" className="gradient-container">
                            <div id="gradient1" className="gradient-part"></div>
                            <div className="legend-value"><p>0.1</p><i>light</i></div>
                        </div>
                        <div id="gradient2container" className="gradient-container">
                            <div id="gradient2" className="gradient-part"></div>
                        </div>
                        <div id="gradient3container" className="gradient-container">
                            <div id="gradient3" className="gradient-part"></div>
                            <div className="legend-value"><p>1</p><i className="innerRainDescription">moderate</i></div>
                        </div>
                        <div id="gradient4container" className="gradient-container">
                            <div id="gradient4" className="gradient-part"></div>
                            <div className="double-text-div">
                                <div id="left-value" className="legend-value"><p>10</p><i className="innerRainDescription">heavy</i></div>
                                <div id="right-value" className="legend-value"><p>&gt;100</p><i>extreme</i></div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }


}