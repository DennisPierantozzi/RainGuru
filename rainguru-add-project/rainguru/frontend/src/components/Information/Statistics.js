import React from 'react';

function Statistics ({displayTooltipLegend}) {

    return(
        <div className='statisticsContents'>
            <div className="legend-Graph">
                <ul className="legend-Graph-list">
                    <li id="labelObservations"> <hr></hr>  Observations</li>
                    <li id="labelPredictions"> <hr></hr>  Predictions</li>
                </ul>
            </div>
            <div className='statistics-box'>
                
                <button id="rmse-explain" className="stats-info" onClick={() => displayTooltipLegend("rmse")}>?</button>
                Rmse: <span id="rmse"></span></div>
            <div className='statistics-box'>
                
                <button id="bias-explain" className="stats-info" onClick={() => displayTooltipLegend("bias")}>?</button>
                Bias: <span id="bias"></span></div>
        </div>
    );
}


export default Statistics;

