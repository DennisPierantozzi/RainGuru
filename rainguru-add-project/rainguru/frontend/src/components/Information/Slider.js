import React from "react";
import Communication from "../Communication"
import AnimationBar from "./AnimationBar";

export default class Slider {

    // stores the latest predictions, slider state (0 to 19), and latitude and longitude of the selected location
    static predictions;
    static sliderValue;
    static lastLat;
    static lastLong;

    /**
     * Retrieves and displays the precipitation data at a specific location.
     * @param latitude the latitude of the specified location.
     * @param longitude the longitude of the specified location.
     */
    static async showPrecipitationData(latitude, longitude) {
        try {
            // update the last lat long value
            Slider.lastLat = latitude;
            Slider.lastLong = longitude;

            // calculate the corresponding rain value stored in the data matrix for the given latitude and longitude
            const y_temp = (latitude - 50.21) / (54.71 - 50.21);
            const x_temp = (longitude - 1.90) / (9.30 - 1.90);
            const y = Math.round(y_temp * 700);
            const x = Math.round(x_temp * 700);
            const precipitation = await Communication.fetchPrecipitation(699 - y, x);

            // find the maximum precipitation amount for the selected location
            const predictions = [];
            let maxRain = 0;
            for (let t = 0; t < 20; t++) {
                if (precipitation[t] > maxRain) {
                    maxRain = precipitation[t];
                }
                predictions.push(precipitation[t]);
            }

            // update the static predictions variable
            Slider.predictions = predictions;

            // dynamically update the interval information and append newly created node to "intervalInfo" div
            const intervalInfoDiv = document.getElementById("intervalInfo");
            intervalInfoDiv.innerHTML = "";
            let timeString = AnimationBar.fullMarks[Slider.sliderValue];
            let intervalInfoDivNode;
            // when precipitation amount is a whole number, add .0 to it
            if (Slider.predictions[Slider.sliderValue] % 1 === 0) {
                intervalInfoDivNode = document.createTextNode(timeString + " - " + Slider.predictions[Slider.sliderValue] + ".0 mm/h");
            } else {
                intervalInfoDivNode = document.createTextNode(timeString + " - " + Slider.predictions[Slider.sliderValue] + " mm/h");
            }
            intervalInfoDiv.appendChild(intervalInfoDivNode);

            // dynamically change maximum rain amount for the given location
            const maxRainDiv = document.getElementById("maxRain");
            if (maxRain % 1 === 0) {
                maxRainDiv.innerHTML = "Max: " + maxRain + ".0 mm/h";
            } else {
                maxRainDiv.innerHTML = "Max: " + maxRain + " mm/h";
            }

            // dynamically set maximum scale value
            const maxScaleDiv = document.getElementById("max-scale");
            let maxScale = maxRain.toFixed(2);
            // when maxRain is less than 0.1, set the maxScale to be 0.1
            if (maxRain <= 0.1) {
                maxScale = 0.1;
            }
            maxScaleDiv.innerHTML = maxScale;

            // dynamically set middle scale value
            const midScaleDiv = document.getElementById("mid-scale");
            let midScaleValue;
            // when maxScale is less than 0.1, set the midScaleValue to be 0.1
            if (maxRain <= 0.1) {
                midScaleValue = 0.05;
            } else {
                midScaleValue = (maxRain / 2).toFixed(2);
            }
            midScaleDiv.innerHTML = midScaleValue;

            // call a helper method to create a precipitation chart
            Slider.makeChart(predictions, maxScale);
        } catch (e) {
            console.log(e);
        }
    }

    /**
     * Creates a precipitation chart (filled line chart) and append it to "chart" div.
     * @param predictions Predictions for the next 100 minutes that need to be graphed
     * @param maxScale Indication of the maximum y-value
     */
    static makeChart(predictions, maxScale) {
        const currentTime = new Date();
        const timeStamps = [];
        for (let i = 0; i < 20; i++) {
            timeStamps.push(i);
        }
        const chart = document.getElementById("chart").getContext("2d");
        const predictionGraph = new Chart(chart, {
            type: "line",
            data: {
                labels: timeStamps,
                datasets: [{
                    fill: true,
                    data: predictions,
                    borderColor: "#1167b1",
                    backgroundColor: "rgba(173, 216, 230, 0.7)"
                }]
            },
            options: {
                maintainAspectRatio: false,
                scales: {
                    xAxes: [{
                        display: false,
                        gridLines: {
                            color: "rgba(0, 0, 0, 0)"
                        }
                    }],
                    yAxes: [{
                        display: false,
                        ticks: {
                            beginAtZero: true,
                            max: maxScale * 1.05
                        },
                        gridLines: {
                            color: "rgba(0, 0, 0, 0)",
                        },
                    }]
                },
                legend:
                    {
                        display: false
                    }
            }
        })
    }
}
