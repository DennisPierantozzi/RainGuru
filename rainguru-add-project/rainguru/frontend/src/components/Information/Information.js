import React, {Component} from "react";
import AnimationBar from "./AnimationBar";
import Legend from "./Legend";

export default class Information extends Component {

    /**
     * Sets the information section in the starting state.
     * @param props contains optional properties of the information section.
     */
    constructor(props) {
        super(props);

        // state variables to pass along to the animation bar
        this.state = {
            animationBarUpdateProp: this.props.animationBarUpdateProp
        }
    }

    /**
     * Changes the state according to the prop to pass the change down to the animation bar.
     *
     * @param nextProps the props given down from the App component.
     * @param prevState the last state of the component before these new props.
     */
    static getDerivedStateFromProps(nextProps, prevState) {
        // if the animation bar update prop changes send it through
        if (nextProps.animationBarUpdateProp !== prevState.animationBarUpdateProp) {
            return {animationBarUpdateProp : nextProps.animationBarUpdateProp};
        }
        else return null;
    }

    /**
     * Renders the content of the information div at the bottom of the screen, so the animation bar and the legend.
     * @returns the rendered content of the information div.
     */
    render() {
        return (<div className="informationContents" data-testid="information">
            <AnimationBar updateProp={this.state.animationBarUpdateProp}/>
            <Legend/>
        </div>);
    }
}
