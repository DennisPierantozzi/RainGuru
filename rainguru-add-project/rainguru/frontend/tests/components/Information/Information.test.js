import React from "react";
import {render, screen} from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import Information from "../../../src/components/Information/Information";
import Communication from "../../../src/components/Communication";

beforeEach(() => {
    // set up communication with a timestamp (23:50), so it can't break the tests since it will be requested in most
    // use a correction for timezones so each timezone runs it with the same local time to keep the tests intact
    const timeZoneOffset = ((new Date()).getTimezoneOffset()) * 60;
    Communication.dataTimestamp = new Date(1000 * (1653263400 + timeZoneOffset));
});

describe("Information tests", () => {
    it("matches snapshot", () => {
        const state = renderer.create(<Information/>).toJSON();
        expect(state).toMatchSnapshot();
    });

    it("renders information", () => {
        render(<Information/>);
        const element = screen.getByTestId("information");
        expect(element).toBeTruthy();
    });


    describe("passing props tests", () => {
        it("no change in update prop", () => {
            expect(Information
                .getDerivedStateFromProps({animationBarUpdateProp: 4}, {animationBarUpdateProp: 4})).toBeNull();
        });

        it("change in update prop", () => {
            expect(Information
                .getDerivedStateFromProps({animationBarUpdateProp: 5}, {animationBarUpdateProp: 4}))
                .toStrictEqual({animationBarUpdateProp: 5});
        });
    });
});
