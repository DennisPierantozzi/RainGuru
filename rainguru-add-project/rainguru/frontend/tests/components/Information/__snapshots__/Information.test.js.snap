// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Information tests matches snapshot 1`] = `
<div
  className="informationContents"
  data-testid="information"
>
  <div
    className="animationContents"
    data-testid="animationBarDiv"
  >
    <div
      className="animationButtonDiv"
      data-testid="animationButtonDiv"
    >
      <button
        className="animationButton"
        data-testid="slowDownButton"
        id="slowDownButton"
        onClick={[Function]}
      >
        <img
          className="ratioImage"
          src="../../../static/images/1x1.png"
        />
        <svg
          className="button-icon"
          fill="currentColor"
          height="70%"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="70%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M35.9 249.3L244 129.1c5.4-3.1 12.3.6 12.3 6.7V251l211.3-121.9c5.4-3.1 12.3.6 12.3 6.7v240.3c0 6.1-6.9 9.8-12.3 6.7L256.4 261v115.2c0 6.1-6.9 9.8-12.3 6.7L35.9 262.7c-5.2-3-5.2-10.4 0-13.4z"
          />
        </svg>
      </button>
      <button
        className="animationButton"
        data-testid="backwardButton"
        id="backwardButton"
        onClick={[Function]}
      >
        <img
          className="ratioImage"
          src="../../../static/images/1x1.png"
        />
        <svg
          className="button-icon"
          fill="currentColor"
          height="70%"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 256 512"
          width="70%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          />
        </svg>
      </button>
      <button
        className="animationButton"
        data-testid="pauseResumeButton"
        id="pauseResumeButton"
        onClick={[Function]}
      >
        <img
          className="ratioImage"
          src="../../../static/images/1x1.png"
        />
        <svg
          className="button-icon"
          fill="currentColor"
          height="70%"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="70%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M199.9 416h-63.8c-4.5 0-8.1-3.6-8.1-8V104c0-4.4 3.6-8 8.1-8h63.8c4.5 0 8.1 3.6 8.1 8v304c0 4.4-3.6 8-8.1 8zM375.9 416h-63.8c-4.5 0-8.1-3.6-8.1-8V104c0-4.4 3.6-8 8.1-8h63.8c4.5 0 8.1 3.6 8.1 8v304c0 4.4-3.6 8-8.1 8z"
          />
        </svg>
      </button>
      <button
        className="animationButton"
        data-testid="forwardButton"
        id="forwardButton"
        onClick={[Function]}
      >
        <img
          className="ratioImage"
          src="../../../static/images/1x1.png"
        />
        <svg
          className="button-icon"
          fill="currentColor"
          height="70%"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 256 512"
          width="70%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          />
        </svg>
      </button>
      <button
        className="animationButton"
        data-testid="speedUpButton"
        id="speedUpButton"
        onClick={[Function]}
      >
        <img
          className="ratioImage"
          src="../../../static/images/1x1.png"
        />
        <svg
          className="button-icon"
          fill="currentColor"
          height="70%"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="70%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M476.1 249.3L268 129.1c-5.4-3.1-12.3.6-12.3 6.7V251L44.3 129.1c-5.4-3.1-12.3.6-12.3 6.7v240.3c0 6.1 6.9 9.8 12.3 6.7L255.6 261v115.2c0 6.1 6.9 9.8 12.3 6.7L476 262.7c5.3-3 5.3-10.4.1-13.4z"
          />
        </svg>
      </button>
    </div>
    <div
      className="dataBar"
    >
      <div
        className="dataBarPart"
        data-testid="speedValue"
        id="speedValue"
      >
        Animation speed: 5
      </div>
      <div
        className="dataBarPart"
        data-testid="intervalInfo"
        id="intervalInfo"
      />
      <div
        className="dataBarPart"
        data-testid="maxRain"
        id="maxRain"
      />
    </div>
    <div
      className="scale-container"
    >
      <div
        data-testid="max-scale"
        id="max-scale"
      />
      <div
        data-testid="mid-scale"
        id="mid-scale"
      />
      <div
        data-testid="min-scale"
        id="min-scale"
      >
        0
      </div>
    </div>
    <div
      className="animationSliderDiv"
    >
      <hr
        className="middle-line"
      />
      <span
        className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root MuiSlider-marked MuiSlider-trackFalse css-1aznpnh-MuiSlider-root"
        data-testid="animationSlider"
        defaultValue={0}
        id="animationSlider"
        onMouseDown={[Function]}
      >
        <span
          className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          className="MuiSlider-track css-pm1ix7-MuiSlider-track"
          style={
            Object {
              "left": "0%",
              "width": "0%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={0}
          style={
            Object {
              "left": "0%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
          data-index={0}
          style={
            Object {
              "left": "0%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={1}
          style={
            Object {
              "left": "5.2631578947368425%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={1}
          style={
            Object {
              "left": "5.2631578947368425%",
            }
          }
        >
          23:55
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={2}
          style={
            Object {
              "left": "10.526315789473685%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={2}
          style={
            Object {
              "left": "10.526315789473685%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={3}
          style={
            Object {
              "left": "15.789473684210526%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={3}
          style={
            Object {
              "left": "15.789473684210526%",
            }
          }
        >
          0:05
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={4}
          style={
            Object {
              "left": "21.05263157894737%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={4}
          style={
            Object {
              "left": "21.05263157894737%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={5}
          style={
            Object {
              "left": "26.31578947368421%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={5}
          style={
            Object {
              "left": "26.31578947368421%",
            }
          }
        >
          0:15
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={6}
          style={
            Object {
              "left": "31.57894736842105%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={6}
          style={
            Object {
              "left": "31.57894736842105%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={7}
          style={
            Object {
              "left": "36.8421052631579%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={7}
          style={
            Object {
              "left": "36.8421052631579%",
            }
          }
        >
          0:25
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={8}
          style={
            Object {
              "left": "42.10526315789474%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={8}
          style={
            Object {
              "left": "42.10526315789474%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={9}
          style={
            Object {
              "left": "47.36842105263158%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={9}
          style={
            Object {
              "left": "47.36842105263158%",
            }
          }
        >
          0:35
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={10}
          style={
            Object {
              "left": "52.63157894736842%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={10}
          style={
            Object {
              "left": "52.63157894736842%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={11}
          style={
            Object {
              "left": "57.89473684210526%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={11}
          style={
            Object {
              "left": "57.89473684210526%",
            }
          }
        >
          0:45
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={12}
          style={
            Object {
              "left": "63.1578947368421%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={12}
          style={
            Object {
              "left": "63.1578947368421%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={13}
          style={
            Object {
              "left": "68.42105263157895%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={13}
          style={
            Object {
              "left": "68.42105263157895%",
            }
          }
        >
          0:55
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={14}
          style={
            Object {
              "left": "73.6842105263158%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={14}
          style={
            Object {
              "left": "73.6842105263158%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={15}
          style={
            Object {
              "left": "78.94736842105263%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={15}
          style={
            Object {
              "left": "78.94736842105263%",
            }
          }
        >
          1:05
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={16}
          style={
            Object {
              "left": "84.21052631578948%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={16}
          style={
            Object {
              "left": "84.21052631578948%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={17}
          style={
            Object {
              "left": "89.47368421052632%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={17}
          style={
            Object {
              "left": "89.47368421052632%",
            }
          }
        >
          1:15
        </span>
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={18}
          style={
            Object {
              "left": "94.73684210526316%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={18}
          style={
            Object {
              "left": "94.73684210526316%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={19}
          style={
            Object {
              "left": "100%",
            }
          }
        />
        <span
          aria-hidden={true}
          className="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index={19}
          style={
            Object {
              "left": "100%",
            }
          }
        />
        <span
          className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
          data-index={0}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "left": "0%",
              "pointerEvents": undefined,
            }
          }
        >
          <input
            aria-label="Precipitation"
            aria-orientation="horizontal"
            aria-valuemax={19}
            aria-valuemin={0}
            aria-valuenow={0}
            data-index={0}
            disabled={false}
            max={19}
            min={0}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            step={1}
            style={
              Object {
                "border": 0,
                "clip": "rect(0 0 0 0)",
                "direction": "ltr",
                "height": "100%",
                "margin": -1,
                "overflow": "hidden",
                "padding": 0,
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "100%",
              }
            }
            type="range"
            value={0}
          />
        </span>
      </span>
      <div
        className="precipitationChart"
        id="chartContainer"
      >
        <canvas
          id="chart"
        />
      </div>
    </div>
  </div>
  <div
    className="legendContents"
    data-testid="legend-box"
  >
    <div
      className="legendTitle"
    >
      Precipitation [mm/h]
    </div>
    <div
      className="legendContainer"
    >
      <div
        className="legend"
      >
        <div
          className="gradient-container"
          id="gradient1container"
        >
          <div
            className="gradient-part"
            id="gradient1"
          />
          <div
            className="legend-value"
          >
            <p>
              0.1
            </p>
            <i>
              light
            </i>
          </div>
        </div>
        <div
          className="gradient-container"
          id="gradient2container"
        >
          <div
            className="gradient-part"
            id="gradient2"
          />
        </div>
        <div
          className="gradient-container"
          id="gradient3container"
        >
          <div
            className="gradient-part"
            id="gradient3"
          />
          <div
            className="legend-value"
          >
            <p>
              1
            </p>
            <i
              className="innerRainDescription"
            >
              moderate
            </i>
          </div>
        </div>
        <div
          className="gradient-container"
          id="gradient4container"
        >
          <div
            className="gradient-part"
            id="gradient4"
          />
          <div
            className="double-text-div"
          >
            <div
              className="legend-value"
              id="left-value"
            >
              <p>
                10
              </p>
              <i
                className="innerRainDescription"
              >
                heavy
              </i>
            </div>
            <div
              className="legend-value"
              id="right-value"
            >
              <p>
                &gt;100
              </p>
              <i>
                extreme
              </i>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
